pkg/comp-functions/functions/buckets/cloudscalebucket/cloudscalebucket.go

```diff
@@ -0,0 +1,141 @@
+package cloudscalebucket
+
+import (
+	"context"
+	"fmt"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	xfnproto "github.com/crossplane/function-sdk-go/proto/v1beta1"
+	appcatv1 "github.com/vshn/appcat/v4/apis/v1"
+	"github.com/vshn/appcat/v4/pkg/comp-functions/functions/buckets/util"
+	"github.com/vshn/appcat/v4/pkg/comp-functions/runtime"
+	cloudscalev1 "github.com/vshn/provider-cloudscale/apis/cloudscale/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	bucketResName = "cloudscale-bucket"
+	userResName   = "cloudscale-user"
+)
+
+// ProvisionCloudscalebucket will create a bucket in cloudscale.
+// This function will leverage provider-cloudscale to deploy proper users
+// alongside the bucket.
+func ProvisionCloudscalebucket(_ context.Context, svc *runtime.ServiceRuntime) *xfnproto.Result {
+
+	bucket := &appcatv1.ObjectBucket{}
+
+	err := svc.GetObservedComposite(bucket)
+	if err != nil {
+		return runtime.NewFatalResult(err)
+	}
+
+	config, ok := svc.Config.Data["providerConfig"]
+	if !ok {
+		return runtime.NewFatalResult(fmt.Errorf("no providerConfig specified"))
+	}
+
+	err = addBucket(svc, bucket, config)
+	if err != nil {
+		return runtime.NewFatalResult(err)
+	}
+
+	err = addUser(svc, bucket, config)
+	if err != nil {
+		return runtime.NewFatalResult(err)
+	}
+
+	svc.SetConnectionDetail("BUCKET_NAME", []byte(bucket.Spec.Parameters.BucketName))
+	svc.SetConnectionDetail("AWS_REGION", []byte(bucket.Spec.Parameters.Region))
+
+	err = populateEndpointConnectionDetails(svc)
+	if err != nil {
+		return runtime.NewFatalResult(err)
+	}
+
+	return nil
+}
+
+func addBucket(svc *runtime.ServiceRuntime, bucket *appcatv1.ObjectBucket, config string) error {
```

Comment:
This function is identical to the `addBucket` function in the `exoscalebucket` package. It would probably make sense to generalize this function.

Reply:
They are not identical, one creates a `cloudscale` bucket and the other creates an `exoscale` bucket.

Reply:
True. Had a closer look and indeed it's not really feasible